### PR TITLE
Fix MySQL Server version auto-update

### PR DESCRIPTION
**Pull request by greedy52**
Partially fix:
- #38948

Changelog: Fixed an issue that the server version of the registered MySQL databases is not automatically updated upon new connections

The MySQL engine can automatically update the MySQL server version upon a new connection. This functionality broke when we started passing database copies to the engines which makes the update ineffective.

### CODE REVIEW

File: lib/srv/db/common/engines.go [hunk 1]
```diff
@@ -117,6 +117,9 @@ type EngineConfig struct {
 	DataDir string
 	// GetUserProvisioner is automatic database users creation handler.
 	GetUserProvisioner func(AutoUsers) *UserProvisioner
+	// UpdateProxiedDatabase finds the proxied database by name and provide a
```
Comment (camscale) [inline lib/srv/db/common/engines.go:120]: "provide a function to update its status" perhaps could be worded as "calls the given function with that database".

But this is a rather generic function - the called function does not need to necessarily "update" anything. The implementation is really "call a function with the database identified by name". I'm not sure what other operations besides "update" makes sense though, so I'll leave it up to you as to whether this field should have a more generic name than "Update..."

File: lib/srv/db/mysql/engine.go [hunk 2]
```diff
@@ -164,9 +164,16 @@ func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Sessio
...
 	serverVersion := serverConn.GetServerVersion()
 	statusVersion := sessionCtx.Database.GetMySQLServerVersion()
-	// Update only when needed
+	// Update only when needed.
+	// Note that sessionCtx.Database is a copy of the database cached by
+	// database service. Call UpdateProxiedDatabase to do the update instead of
+	// setting the copy.
 	if serverVersion != "" && serverVersion != statusVersion {
-		sessionCtx.Database.SetMySQLServerVersion(serverVersion)
+		return trace.Wrap(e.UpdateProxiedDatabase(sessionCtx.Database.GetName(), func(db types.Database) error {
```
Comment (camscale) [inline lib/srv/db/mysql/engine.go:172]: I'm not a fan of this level of nesting - the happy path is rather hidden. I find Go code to be quite "scannable" if it does not pack too much into one line / function call. Perhaps this?
```
err := e.UpdateProxiedDatabase(sessionCtx.Database.GetName(), func(db types.Database) error {
    db.SetMySQLServerVersion(serverVersion)
    e.Log.WithField("server-version", serverVersion).Debug("Updated MySQL server version.")
    return nil
})
if err != nil {
    return trace.Wrap(err)
}
```

and let the happy path fall through to the end of the function? This also helps if something needs to be added after this block later - the inside of the block does not need to be altered as the happy path continues after it.

File: lib/srv/db/server.go [hunk 3]
```diff
@@ -1104,6 +1104,16 @@ func (s *Server) createEngine(sessionCtx *common.Session, audit common.Audit) (c
 				Clock:      s.cfg.Clock,
 			}
 		},
+		UpdateProxiedDatabase: func(name string, doUpdate func(types.Database) error) error {
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			for _, db := range s.proxiedDatabases {
```
[the remark this answers is not in the document]
Comment (greedy52) [reply lib/srv/db/server.go]: good catch. I copied from the other function and didn't give it too much thought. I updated the other function now as well.

File: lib/srv/db/common/engines.go (hunk 1 above)
Comment (greedy52) [reply lib/srv/db/common/engines.go]: I updated the description a little bit. I didn't come up with a better name so i will leave that as it is unless someone has better options.

File: lib/srv/db/mysql/engine.go (hunk 2 above)
Comment (greedy52) [reply lib/srv/db/mysql/engine.go]: refactored the function a little bit to make it more readable.

**Review by Tener — APPROVED** [review]

File: lib/srv/db/access_test.go [hunk 4]
```diff
@@ -333,6 +343,26 @@ func TestAccessMySQL(t *testing.T) {
...
+		require.NoError(t, err)
+		defer mysqlConn.Close()
+		_, err = mysqlConn.Execute("select 1")
+		require.NoError(t, err)
+
+		// Check if proxied database is updated.
+		updatedDB, err := testCtx.server.getProxiedDatabase("version-update-test")
+		require.NoError(t, err)
+		require.Equal(t, "8.8.8-after", updatedDB.GetMySQLServerVersion())
+	})
```
Comment (smallinsky) [inline lib/srv/db/access_test.go:365]: nit: 
Can you move this into a sperate  function test like `TestServerVersionUpgradeOnConnection(t *testing.t)` 

Currently the TestAccessMySQL  test case is  using table tests approach and adding  "server version update on connection" after table tests is a bit unreadable.

**Review by smallinsky — APPROVED** [review]

File: lib/srv/db/access_test.go (hunk 4 above)
Comment (greedy52) [reply lib/srv/db/access_test.go]: thanks! moved!

**Pull request merged**